### ISSUE TITLE
angular/ng_controller_as_vm: false positive in nested scopes inside controllers

### ISSUE THREAD

**Issue by zachlysobey** [Gillespie59/eslint-plugin-angular#144]
I'm getting the following error in my controller. I *am* using `vm` as suggested, but some of my methods reference (their own) `this`.

> error  You should not use "this" directly. Instead, assign it to a variable called "vm"  angular/ng_controller_as_vm

```javascript
vm.handleDragStart = function(event) {
  this.style.opacity = '0.4';
  vm.dragging = true;
};
```

Here, `this` is *not* referencing the viewModel, and as such shouldn't trigger that error.

**Comment by zachlysobey** [comment]
Not sure if a pull request with a failing test verifying the issue is proper etiquette, I'm new to contributing to Open Source projects. I'll see if I can take time later to see if I can enact the fix myself, but I might not be very effective at it.

**Comment by Gillespie59** [comment]
Perfect! ;)

**Comment by Gillespie59** [comment]
I do not know how can solve this issue. in fact your use case is quiet hard to detect :s

**Comment by tilmanpotthof** [comment]
Although the use of `this` may be useful in this case, it is quite confusing because you can not know what `this` is when you define a function on your controller. Even if we find a way to detect this pattern, it is more likely to miss potential errors than false positives as in your case.

My recommendation is to either find another way without `this` (pass a parameter) or to disable the rule for this line. There are rule and there are exceptions...

    // eslint-disable-line angular/controller-as-vm

**Comment by tilmanpotthof** [comment]
@zachlysobey You could also pass your style object as parameter.

    vm.handleDragStart = function(event, style) {
      style.opacity = '0.4';
      vm.dragging = true;
    };

@Gillespie59 I close this issue and the PR #146. I see still no reasonable way to differentiate between both patterns.
